### PR TITLE
Improved search feature for API docs

### LINKED ISSUE #1792: [API doc browser] Method index/search
There should be a searchable index of all methods

### PR DESCRIPTION
**Pull request by straight-shoota**
This PR greatly improves client-side search on generated HTML pages of the API docs.
It uses the program description available as JSON data from #4746 and implements a custom search algorithm in JavaScript.
The search results include types, constants, methods, macros matching the query string in one or more of the fields name, namespace, description, arguments or type.
A custom ranking algorithm sorts results based on matching quality and relevance.

The search-algorithm is actually quite simple, it uses no search index or advanced information retrieval functions. It simply traverses through all the types and tries to find matching features. This is really just fast enough.
I considered using something like lunr.js but this would have been complicated to setup to create a domain-specific search experience. With the custom algorithm it is very easy to fine tune matching and ranking behaviour to allow developers to find what they need as fast as possible.

This approach has some drawbacks with regards to fulltext search capabilities (especially concerning docstrings), but this can certainly be improved upon.

## Search features:
* prefix filter: `#` matches instance methods, `.` matches class methods and macros, `:`/`::` matches namespace
* search is generally case-insensitive but results with case-equality areranked higher.
* search results list can be navigated by key (when in focus):
  * `i`, `c`, `ArrowUp`: move upwards
  * `j`, `h`, `ArrowDown`: move downwards
  * `Enter`: load result
  * `Escape`: cancel search
  * arrows, `Enter` and `Escape` also work when the search-input is in focus
* support for URL fragment-based search query: `/#q=mysearch` will initialize a search at page load
* page-wide shortkeys to start search: `s` or `/`

## Preview
[Online Preview](https://straight-shoota.github.io/crystal/docs-search-only/)

This is intended to be alongside #4755 which makes it look even better ([Online Preview](https://straight-shoota.github.io/crystal/doc-search/)) though it can be merged separately.

Closes #1792

### CODE REVIEW

File: src/compiler/crystal/tools/doc/generator.cr [hunk 1]
```diff
@@ -78,6 +78,7 @@ class Crystal::Doc::Generator
 
     main_index = Main.new(raw_body, Type.new(self, @program), repository_name)
     File.write File.join(@output_dir, "index.json"), main_index
+    File.write File.join(@output_dir, "index.jsonp"), main_index.to_jsonp
```
Comment (MakeNowJust) [inline src/compiler/crystal/tools/doc/generator.cr]: I think `search-index.js` (from Rust doc) is better than `index.jsonp`. The extension `jsonp` is unsound.

Comment (Sija) [reply src/compiler/crystal/tools/doc/generator.cr]: Yeah, why JSONP? Still, there might a problem with loading `json` files over `file:///`.

Comment (straight-shoota) [reply src/compiler/crystal/tools/doc/generator.cr]: `search-index.js` seems like a good idea 👍

**Comment by straight-shoota** [comment]
@Sija Yes, browsers can't simply load arbitrary files over `file` protocol. That's why we need to load the search index as a javascript file to enable search when reading the docs locally.

We could also encourage people to load the docs through a local HTTP server like [serve](https://github.com/SuperPaintman/serve/) (or at best include one directly into the docs generator) but this is always one step more complicated and the workaround with JSONP is super easy.
`search-index.js` can obviously be ignored when deploying to a HTTP server.

**Comment by RX14** [comment]
Why isn't it `j`/`k` for up/down? That's the standard vim controls...

**Comment by RX14** [comment]
Bug reports:
- pressing `/` only works the first time after page load on firefox, because the second time it brings up quick search.
- If i type too quickly that search doesn't complete before i press `enter`, (i.e. pressing `string<enter>` with no delay) it doesn't wait until the search is complete before jumping.
- pressing enter to confirm a search doesn't clear the search box and reset the element focus.

**Comment by straight-shoota** [comment]
Thanks for the testing.

* I copies the key bindings from api.rubyonrails.org - seems like they had it wrong and I didn't notice. Will change to `j/k`
* I've only done rudimentary tests in firefox. Seems a bit odd. I noticed that it might sometimes be a focus issue.
* Fast jumping will be fixed.
* This is supposed to be a feature. Additionally, also the sidebar scroll viewport stays the same. I find it useful this way, so you can easily look at a different search result instead of having to start over.

**Comment by RX14** [comment]
The solution is to define a peek button p which does the same as enter but focusses the search bar back again. You could even add ctrl-p so you don't need to press the down arrow

**Comment by straight-shoota** [comment]
@RX14 I'm not sure, I can follow.
In which state should that peek button `p` be available? It can't be while you're typing. Needs a modifier like `Ctrl`. What's the problem with the current behaviour?

**Comment by RX14** [comment]
@straight-shoota the problem is that it's unexpected. When i'm at the docs my muscle memory is to do `s<query>`. If I have to evaluate the page to see exactly what element im focussed on it's just going to make me a new hundred milliseconds longer because I have to change how I search. (yes I do care about it that much)

The rule should be that after I press enter I should be exactly the same node of the search state machine as I am when the page loads. And the starting node shouldn't depend on anything, a refresh should be a completely blank slate.

For people who want your behaviour of keeping the search results to browse through them, they should have a "peek" shortcut (ctrl-p) to peek the results of a search result, but they're still at the search list so they can find another result if it's incorrect. I'll probably never use peek. 99% of the time i've arrived in the right file, or at the least i'll need to correct my search term not just move to the second result. In fact i'll rarely be moving to the second result, largely my intended result will be rank 1, or i'll be refining my search term.

**Comment by RX14** [comment]
Another thing which would greatly help for muscle memory would be making it so that pressing `s` highlights the results of the search box. This is a natural way to make it so that using `s<query>` never appends to an existing search query, it's always a brand new search. If you want to edit your existing query, use right arrow to get to the end of the query, or use left arrow to add a prefix. This is much more natural and much easier to get to the end/start of the search query than just changing the focus but not selecting the whole contents.

**Comment by straight-shoota** [comment]
If I understand you correctly, with "Highlight the results of the search box" you mean "select the text in the query input field"?

**Comment by RX14** [comment]
@straight-shoota yep.

**Comment by RX14** [comment]
@straight-shoota have you had any more time to work on this?

**Comment by straight-shoota** [comment]
@RX14 I've implemented that the results list is not automatically reloaded on a new page and when you focus the search field, the content gets selected.

I haven't done the peek feature yet. `Strg + P` won't work because in a browser this is the shortcut for printing. I'm not sure if it is worth to implement at all and if so, which shortcut would be suited for this.
Maybe `Shift + Enter`... but probably it's not necessary at all.

**Comment by RX14** [comment]
@straight-shoota bug: typing `result` types `re` selects the whole entry field, then overwrites it with `ult`. I'm pretty sure *that's* unintentional.

**Comment by straight-shoota** [comment]
Woah, yep. That event shouldn't trigger when the input is selected.

**Comment by straight-shoota** [comment]
I've fixed this and added a modal window listing shortkeys (can be opened with `?`).

**Comment by faustinoaq** [comment]
Awesome improvement to documentation 🎉 

Would be possible to hide sidebar on small screens and  add a hamburger button?

**Comment by straight-shoota** [comment]
@faustinoaq Yes, this is the next step after #4755

**Comment by RX14** [comment]
I think that the input box should be explicitly cleared and unfocussed at page load. Firefox seems to restore the textbox content and focus when the page is refreshed.

**Comment by straight-shoota** [comment]
@RX14 This is actually done on purpose. Each search query is stored in session storage and the last value is retrieved on page load. This is intended to help refine a query so you don't have to type again.
Autofocus on page load is also a nice feature, because when you open a docs page, the only thing you'd want to type is a search query. This is a common feature, witnessed for example on https://api.rubyonrails.com 

But I see that the combination of both can be confusing, and maybe we could remove the auto-filling but keep the focus.

**Comment by RX14** [comment]
@straight-shoota autofilling is actually OK, since `s` selects what's filled anyway. Changing focus on pageload isnt.

**Comment by straight-shoota** [comment]
Being mostly a mouse user I didn't put that into consideration.
I guess it comes down to mouse users having to press `s` or keyboard users having to press `Esc` 🤣 But since the former is just a nice shortcut the latter is truly a huge restriction for usability, it's certainly better not to have autofocus.
Plus, this makes it possible to keep the autofill.

**Comment by RX14** [comment]
@straight-shoota have you updated the preview site? I still get the focus on the search box.

**Comment by straight-shoota** [comment]
Oh, sry. Now it is updated.

**Comment by straight-shoota** [comment]
Fixed it, preview is now correct.

**Comment by RX14** [comment]
@straight-shoota the tiniest nit: seems that doing `sstring<enter>` fast enough still doesn't always search properly. Inputs should be buffered up and all responded to as if the search had been instant.

**Comment by RX14** [comment]
But apart from that this feels fantastic.

**Comment by RX14** [comment]
you probably shouldn't search for every inputted character if the user types fast to try and minimise search lag. Ideally, inputting another character should immediately cancel the previous search, but not sure how easy that is in JS.

**Comment by straight-shoota** [comment]
I think I got this issue fixed. If you're pressing enter while a search is in progress, the action is delayed until the search has finished.

There is already a debounce in place in `performSearch()` - it has been around even before this PR. It actually doesn't abort a search but only starts a search if there has not been another key event for 200 ms. This is a fairly common implementation for such a feature in JavaScript. It would be difficult to simply abort a search that has already begun.

**Comment by RX14** [comment]
It's almost perfect!

More bugs:
- using the arrow keys to select another result before pressing enter breaks navigation.
- using the arrow keys don't keep the selected search result on screen, it should scroll the results box, not only to keep the selected result on the page, but it should keep the selected result in (or close to) the center of the screen. This way I can see context of what i'm scrolling towards instead of it being off the screen.

**Comment by straight-shoota** [comment]
This seems odd, no. 1 should work and it does on my local copy. I've reupload it, and now it looks fine. Maybe I should have some automation for this...

No. 2 is thankfully quite easy to solve, the DOM API already implements this behaviour.

**Comment by RX14** [comment]
Any update on this?

**Comment by straight-shoota** [comment]
I think I have addressed your latest issues (in code and the online preview) and am awaiting a review. Let me know if something else needs to be done.

One thing I have noticed is the ranking algorithm sometimes does not exactly do what one would expect... It is however not quite easy to debug.
I'm thinking if it would maybe better to switch to an index based ranking instead of the current direct comparison between results. But this could also be worked out after this PR gets merged.
I think the search results are *good enough* and certainly a huge improvement.

**Comment by RX14** [comment]
@straight-shoota when i open up the search results, and hold the down arrow, there's definitely nothing scrolling in firefox 57.

Also, the animation for the seclector means that when you're holding down the down arrow the selector is hardly visible.

**Comment by straight-shoota** [comment]
It turns out that cool and easy DOM feature for scroll alignment is only experimental and not supported by the newest Firefox. So I had to implement this directly. It's not as neat as with `scrollIntoView + options` but I think it will do.

You can't read anything anyway when you're holding down the arrow. To make this better, we could slow down the movement - but that would defeat the purpose of this fast-scrolling feature.

**Comment by RX14** [comment]
This is absolutely fantastic.

**Comment by faustinoaq** [comment]
screenshot_20171204_011609

Looks like README need some `margin/padding`, but I guess it will change on #4755 

BTW, How Can We add other files like README (guides, pages, etc) using `crystal doc`, something like hexdocs:

screenshot_20171204_011923

Would be very useful to have normal documentation (PAGES), and api documentation in the same sidebar 😄 just like hexdocs 👍

**Comment by straight-shoota** [comment]
Can we get this merged before Christmas? Would be a nice present for many Crystal developers... =) 🎄 🎁

**Comment by monouser7dig** [comment]
Apart from further refinements this already is a giant leap forward in terms of usability!
Thanks @straight-shoota 

Would like to see this merged as well, in the meantime I use your link.

**Comment by RX14** [comment]
🎉 🎉 🎉 I've been waiting for this forever

**Comment by straight-shoota** [comment]
@RX14 The current docs could already be upgraded to a version built by a compiler including this PR.

**Comment by straight-shoota** [comment]
In fact, I would really appreciate it if it could be updated soon. That would allow us to catch and fix errors (specifically improvements for the result ranking) which will only emerge if it is broadly used. This way we could get fixes before the next release.

**Comment by RX14** [comment]
I'd be okay with this, not sure about the rest of the core team @straight-shoota.

**Comment by asterite** [comment]
Using better docs for the current version sounds good to me.

**Pull request merged**